transformations/butter_fingers_perturbation_for_IL/transformation.py

```diff
@@ -0,0 +1,110 @@
+import os
+import itertools
+import random
+import json
+
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+'''
+    This transformation adds noise to all types of text sources (sentence, paragraph, etc.), which is proportional to noise erupting from keyboard typos resulting in common spelling errors. 
+    We have expanded the existing implementation for English (source) to a few Indian Languages:
+    1. Bangla ('bn'), 
+    2. Gujarati ('gu'), 
+    3. Hindi ('hi'), 
+    4. Kannada ('kn'), 
+    5. Malayalam ('ml'), 
+    6. Oriya ('or'), 
+    7. Punjabi ('pa'), 
+    8. Tamil ('ta'), 
+    9. Telugu ('te')
+
+    args:
+        language (str): base language for butter fingers
+        keyboard (str): preffered keyboard schema (currently available: "inscript")
+        seed (int, default = 0): seed for reproducibility 
+'''
+
+def butter_finger(text, key_approx, prob=0.1,  seed=0, max_outputs=1):
+
+    random.seed(seed)
+    
+    prob_of_typo = int(prob * 100)
+    perturbed_texts = []
+    for _ in itertools.repeat(None, max_outputs):
+        butter_text = ""
+        for letter in text:
+            if letter not in key_approx.keys():
+                new_letter = letter
+            else:
+                if random.choice(range(0, 100)) <= prob_of_typo:
+                    new_letter = random.choice(key_approx[letter])
+                else:
+                    new_letter = letter
+
+            butter_text += new_letter
+        perturbed_texts.append(butter_text)
+    return perturbed_texts
+
+
+"""
+Butter Finger implementation borrowed and edited from https://github.com/alexyorke/butter-fingers.
+"""
+
+
+class ButterFingersPerturbationForIL(SentenceOperation):
+    tasks = [
+        TaskType.TEXT_CLASSIFICATION,
+        TaskType.TEXT_TO_TEXT_GENERATION,
+        TaskType.TEXT_TAGGING,
+    ]
+
+    keywords = ["morphological", "noise", "rule-based", "high-coverage", "high-precision",
+                "unnatural-sounding", "unnaturally-written", "high-generations"]
```

Comment:
Butter fingers would indeed mean they are naturally-written right? You can skip unnatural-sounding...

Reply:
Understood. I'll make the change.